### PR TITLE
Project Setup; and Person, Student and Teacher Classes Completed

### PR DESCRIPTION
**Pull request by ometman**
I have completed the creation of the initial classes, Person, Student and Teacher on branch 'initial-classes':

** The Person Class has the following:** 
  - Instance vars: @id, @name, and @age.
  - Constructor with name, age, and parent_permission as parameter. 
     -  name and parent_permission are optional and have default values of "Unknown" and true.
  - Getters: @id, @name, and @age.
  - Setters: @name and @age.
  - A Private method of_age? that returns true if @age is greater or equal to 18 and false otherwise.
  - A Public method can_use_services? that returns true if person is of age or if they have permission from parents.

**The Student class has the following:**
  - Inherits from Person.
  - A constructor which extends parent's constructor by adding @classroom and a parameter for it.
  - Method play_hooky that returns "¯\(ツ)/¯".

**The Teacher class has the following:**
  - Inherits from Person.
  - A constructor which extends parent's constructor by adding @specialization and a parameter for it.
  - Override can_use_services? so it always returns true.

### CODE REVIEW

File: person_class.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  private
+
+  def of_age?
+    @age >= 18
+  end
+
+  public :can_use_services
+
+  def can_use_services?
+    of_age? || parent_permission
```
Comment (TedLivist) [inline person_class.rb]: - [ ] This should be `@parent_permission`, not `parent_permission`

File: person_class.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    @parent_permission = parent_permission
+  end
+
+  private
+
+  def of_age?
+    @age >= 18
+  end
+
+  public :can_use_services
```
Comment (TedLivist) [inline person_class.rb]: - [ ] There is no need specifying this `can_use_service` here. Additionally it is not the correct name. Just remove the `:can_use_services` and leave the `public`

File: person_class.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class Person
+  attr_accessor :id, :name, :age
+
+  def initialize(id, age, name = 'UNKNOWN', parent_permission: true)
+    @id = id
```
Comment (TedLivist) [inline person_class.rb]: - [ ] **[OPTIONAL]** We can generate the `@id` so we don't need to pass it.

  ```
  def initialize(age, name = 'UNKNOWN', parent_permission: true)
    @id = rand(1...1000)
  ```

File: student_class.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+require_relative 'person_class'
+
+class Student < Person
+  attr_reader :classroom
+
+  def initialize(id, age, classroom, name = 'UNKNOWN', parent_permission: true)
+    super(id, name, age, parent_permission)
```
Comment (TedLivist) [inline student_class.rb]: - [ ] Kindly use the correct arrangement of parameters in the `super` method. `age` should come before `name`, and `parent_permission` should be assigned correctly.

  ```
  super(id, age, name, parent_permission: parent_permission)
  ```

File: teacher_class.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+require_relative 'person_class'
+
+class Teacher < Person
+  attr_reader :specialization
+
+  def initialize(id, age, specialization, name = 'UNKNOWN')
+    super(id, name, age)
```
Comment (TedLivist) [inline teacher_class.rb]: - [ ] Kindly use the correct arrangement of parameters in the `super` method. `age` should come before `name`.

  ```
  super(id, age, name)
  ```

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @ometman,

Good job so far!
✔️ Implementation of classes in different files based on convention
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person_class.rb (hunk 2 above)
Comment (ometman) [reply person_class.rb]: correct name?

File: person_class.rb (hunk 3 above)
Comment (ometman) [reply person_class.rb]: yeah, some special id gen. can be used.

File: student_class.rb (hunk 4 above)
Comment (ometman) [reply student_class.rb]: whoosh, parameter order mismatch. thanks

File: person_class.rb (hunk 2 above)
Comment (TedLivist) [reply person_class.rb]: Every method under the public keyword is public so there's no need specifying the method here. Also you're missing '?' in the name there.

File: student_class.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+require_relative 'person_class'
+
+class Student < Person
+  attr_reader :classroom
+
+  def initialize(id, age, classroom, name = 'UNKNOWN', parent_permission: true)
```
Comment (zeff96) [inline student_class.rb]: - [ OPTIONAL ] Hey! 👋 In the Person class, I noticed that the id parameter is no longer being passed. Consequently, there's no need to inherit it in the superclass of the Student class. 🚫🆔 Let's streamline the code and remove any unnecessary parameters for clarity and efficiency. Thanks for optimizing! 👍

```suggestion
class Student < Person 
  def initialize(age, classroom, name = "Unknown", parent_permission = true)  
    super(age, name, parent_permission)
    @classroom = classroom
  end
  def play_hooky
    "¯\(ツ)/¯"
  end
end
```

File: teacher_class.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+require_relative 'person_class'
+
+class Teacher < Person
+  attr_reader :specialization
+
+  def initialize(id, age, specialization, name = 'UNKNOWN')
```
Comment (zeff96) [inline teacher_class.rb]: - [ OPTIONAL] Hey! 👋 In the Person class, I noticed that the id parameter is no longer being passed. Consequently, there's no need to inherit it in the superclass of the Teacher class. 🚫🆔 Let's streamline the code and remove any unnecessary parameters for clarity and efficiency. Thanks for optimizing! 👍 

```suggestion
class Teacher < Person
  def initialize(age, specialization, name = "Unknown", parent_permission = true)
    super(age, name, parent_permission)
    @specialization = specialization
  end
  
  def can_use_services?
    true
  end
end
```

**Review by zeff96 — APPROVED** [review]
### Status: APPROVED :green_circle: 
Hi @ometman , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

congrats

## Highlights
:heavy_check_mark: Linters passing
:heavy_check_mark: Correct GitFlow

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: teacher_class.rb (hunk 7 above)
Comment (ometman) [reply teacher_class.rb]: Yeah, Good review. 😄

**Pull request merged**